File: Sample/NendSDK_Sample/NativeAdCarouselCell.m
//
//  NativeAdCarouseCell.m
//  NendSDK_Sample
//
//  Copyright © 2016年 F@N Communications. All rights reserved.
//

#import "NativeAdCarouselCell.h"
#import <NendAd/NADNativeClient.h>
#import "NativeAdCarouselView.h"

#define cellWidth       [UIScreen mainScreen].bounds.size.width
#define timerInterval     3.0f

dispatch_source_t CreateTimer(double interval, dispatch_queue_t queue, dispatch_block_t block)
{
    dispatch_source_t timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue);
    dispatch_source_set_timer(timer, dispatch_time(DISPATCH_TIME_NOW, interval * NSEC_PER_SEC), interval * NSEC_PER_SEC, (1ull * NSEC_PER_SEC) / 10);
    dispatch_source_set_event_handler(timer, block);
    dispatch_resume(timer);
    
    return timer;
}

static const int adCount = 5; // 最大5枚
static const float adPortraitWidth = 320.f; // 縦向き　広告横幅
static const float adPortraitHeight = 325.f; // 縦向き 広告高さ
static const float adLandscapeWidth = 580.f; // 横向き　広告横幅
static const float adLandscapeHeight = 200.f; // 横向き　広告高さ

@interface NativeAdCarouselCell ()<NADNativeDelegate, UIScrollViewDelegate>

@property (nonatomic) NADNativeClient *client;
@property (nonatomic) NSMutableArray *adViewsP;
@property (nonatomic) NSMutableArray *adViewsL;
@property (nonatomic) NSMutableArray *ads;
@property (nonatomic) UIScrollView *scrollView;

@property (nonatomic) float adLandscapeWidth; // 横向き　広告横幅
@property (nonatomic) dispatch_source_t timerP;
@property (nonatomic) float pointP;
@property (nonatomic) int pageP;
@property (nonatomic) dispatch_source_t timerL;
@property (nonatomic) float pointL;
@property (nonatomic) int pageL;

@end

@implementation NativeAdCarouselCell

- (void)awakeFromNib {
    // Initialization code
    [super awakeFromNib];
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];
    // Configure the view for the selected state
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    
    self.scrollView = [[UIScrollView alloc] init];
    self.scrollView.delegate = self;
    self.scrollView.backgroundColor = [UIColor clearColor];
    self.scrollView.showsHorizontalScrollIndicator = NO;
    self.scrollView.directionalLockEnabled = YES;
    self.scrollView.pagingEnabled = NO;
    self.scrollView.bounces = NO;
    self.scrollView.decelerationRate = 0.3;
    [self addSubview:self.scrollView];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(layoutUpdate:) name:@"layoutUpdate" object:nil];
    
    // 自動スクロール
    self.pointP = 0;
    self.pageP = 1;
    self.pointL = 0;
    self.pageL = 1;
    
    return self;
}

- (void) initAd {
    if (self) {
        self.client = [[NADNativeClient alloc] initWithSpotId:@"485504" apiKey:@"30fda4b3386e793a14b27bedb4dcd29f03d638e5"];
        self.ads = [NSMutableArray array];
        self.adViewsP = [NSMutableArray array];
        self.adViewsL = [NSMutableArray array];
        
        // 5ページ分の広告を先にまとめて取得　縦画面
        [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
        __weak typeof(self) weakSelf = self;
        dispatch_group_t group = dispatch_group_create();
        dispatch_apply(adCount, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(size_t i) {
            dispatch_group_enter(group);
            [self.client loadWithCompletionBlock:^(NADNative *ad, NSError *error) {
                if (ad) {
                    ad.delegate = weakSelf;
                    UINib *nibP = [UINib nibWithNibName:@"NativeAdCarouselPortraitView" bundle:nil];
                    UIView<NADNativeViewRendering> *viewP = [nibP instantiateWithOwner:nil options:nil][0];
                    UINib *nibL = [UINib nibWithNibName:@"NativeAdCarouselLandscapeView" bundle:nil];
                    UIView<NADNativeViewRendering> *viewL = [nibL instantiateWithOwner:nil options:nil][0];
                    
                    [weakSelf.ads addObject:ad];
                    [weakSelf.adViewsP addObject:viewP];
                    [weakSelf.adViewsL addObject:viewL];
                } else {
                    NSLog(@"%@", error);
                }
                dispatch_group_leave(group);
            }];
        });
        
        dispatch_group_notify(group, dispatch_get_main_queue(), ^{
            [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
            // 広告ビューを描画
            [self setAd];
        });
    }
}

#pragma mark UIScrollView

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate {
    [self animation];
}

-(void)scrollViewDidEndDecelerating:(UIScrollView *)_scrollView{
    [self animation];
}

- (void) animation {
    int distance = self.scrollView.contentOffset.x;
    float adWidth;
    if ((self.direction).integerValue == 1) {
        adWidth = adPortraitWidth;
        
        for (int i = 1; i <= adCount; i ++) {
            if (distance + cellWidth / 2 < adWidth * i) {
                if (i == 1) {
                    distance = 0;
                } else if (i == adCount) {
                    distance = adWidth * i - cellWidth;
                } else {
                    distance = adWidth * i - adWidth / 2 - cellWidth / 2;
                }
                self.pageP = i;
                break;
            }
        }
        
        self.pointP = distance;
    } else {
        adWidth = self.adLandscapeWidth;
        
        for (int i = 1; i <= adCount; i ++) {
            if (distance + cellWidth / 2 < adWidth * i) {
                if (i == 1) {
                    distance = 0;
                } else if (i == adCount) {
                    distance = adWidth * i - cellWidth;
                } else {
                    distance = adWidth * i - adWidth / 2 - cellWidth / 2;
                }
                self.pageL = i;
                break;
            }
        }
        
        self.pointL = distance;
    }
    
    [self.scrollView setContentOffset:CGPointMake(distance, 0) animated:YES];
}

- (void) layoutUpdate:(NSNotification*) notification {
    self.direction = notification.object;
    
    for (UIView *subview in self.scrollView.subviews) {
        [subview removeFromSuperview];
    }
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self setAd];
    });
}

-(void) setAd {
    if ((self.direction).intValue == 1) {
        for (int i = 0; i < self.adViewsP.count; i ++) {
            NativeAdCarouselView *viewP = (self.adViewsP)[i];
            viewP.frame = CGRectMake(i * adPortraitWidth, 0, adPortraitWidth, adPortraitHeight);
            
            dispatch_async(dispatch_get_main_queue(), ^{
                NADNative *ad = (self.ads)[i];
                [ad intoView:(UIView<NADNativeViewRendering> *)viewP advertisingExplicitly:NADNativeAdvertisingExplicitlyPR];
            });
        }
    } else if ((self.direction).intValue == 2) {
        for (int i = 0; i < self.adViewsL.count; i ++) {
            NativeAdCarouselView *viewL = (self.adViewsL)[i];
            viewL.frame = CGRectMake(i * adLandscapeWidth, 0, adLandscapeWidth, adLandscapeHeight);
            
            dispatch_async(dispatch_get_main_queue(), ^{
                NADNative *ad = (self.ads)[i];
                [ad intoView:(UIView<NADNativeViewRendering> *)viewL advertisingExplicitly:NADNativeAdvertisingExplicitlyPR];
            });
        }
    }
    
    // スクロールビューに広告ビューを格納
    [self layoutUpdate];
}

- (void) layoutUpdate {
    float width;
    float height;
    
    if ((self.direction).integerValue == 1) {
        self.adLandscapeWidth = adPortraitWidth;
        width = self.adLandscapeWidth;
        height = adPortraitHeight;
        
        for (int i = 0; i < self.adViewsP.count; i ++) {
            NativeAdCarouselView *adViewP = (self.adViewsP)[i];
            adViewP.index = i;
            [adViewP frameUpdate:self.direction];
            [self.scrollView addSubview:adViewP];
        }
        [self.scrollView setContentOffset:CGPointMake(self.pointP, 0) animated:YES];
    } else {
        if (cellWidth < adLandscapeWidth) {
            self.adLandscapeWidth = cellWidth - 20;
        } else {
            self.adLandscapeWidth = adLandscapeWidth;
        }
        width = self.adLandscapeWidth;
        height = adLandscapeHeight;
        
        for (int i = 0; i < self.adViewsL.count; i ++) {
            NativeAdCarouselView *adViewL = (self.adViewsL)[i];
            adViewL.index = i;
            [adViewL frameUpdate:self.direction];
            [self.scrollView addSubview:adViewL];
        }
        [self.scrollView setContentOffset:CGPointMake(self.pointL, 0) animated:YES];
    }
    
    self.scrollView.frame = CGRectMake(0.f, 0.f, cellWidth, height);
    self.scrollView.contentSize = CGSizeMake(width * adCount, 0);
    
    self.pageP = 1;
    self.pageL = 1;
    self.pointP = 0;
    self.pointL = 0;
    [self.scrollView setContentOffset:CGPointMake(0, 0) animated:NO];
    // 自動スクロールタイマー設置
    if (adCount > 1) {
        [self setTimer];
    }
}

-(void) setTimer {
    [self cancelTimer];
    
    __weak typeof(self) weakSelf = self;
    dispatch_queue_t queue = dispatch_get_main_queue();
    
    if ((self.direction).intValue == 1) {
        self.timerP = CreateTimer(timerInterval, queue, ^{
            [weakSelf move];
        });
    } else if ((self.direction).intValue == 2) {
        self.timerL = CreateTimer(timerInterval, queue, ^{
            [weakSelf move];
        });
    }
}

- (void)cancelTimer
{
    if (self.timerP) {
        dispatch_source_cancel(self.timerP);
        self.timerP = nil;
    }
    if (self.timerL) {
        dispatch_source_cancel(self.timerL);
        self.timerL = nil;
    }
}

- (void)move
{
    if ((self.direction).intValue == 1) {
        if (self.pageP == adCount - 1){
            self.pointP = adPortraitWidth * adCount - cellWidth;
            self.pageP ++;
            [self.scrollView setContentOffset:CGPointMake(self.pointP, 0) animated:YES];
        } else if (self.pageP == 1) {
            self.pointP += adPortraitWidth * 1.5 - cellWidth / 2;
            self.pageP ++;
            [self.scrollView setContentOffset:CGPointMake(self.pointP, 0) animated:YES];
        } else if (self.pageP == adCount){
            self.pointP = 0;
            self.pageP = 1;
            [self.scrollView setContentOffset:CGPointMake(0, 0) animated:NO];
        } else {
            self.pointP += adPortraitWidth;
            self.pageP ++;
            [self.scrollView setContentOffset:CGPointMake(self.pointP, 0) animated:YES];
        }
        
    } else if ((self.direction).intValue == 2) {
        if (self.pageL == adCount - 1){
            self.pointL = self.adLandscapeWidth * adCount - cellWidth;
            self.pageL ++;
            [self.scrollView setContentOffset:CGPointMake(self.pointL, 0) animated:YES];
        } else if (self.pageL == 1) {
            self.pointL += self.adLandscapeWidth * 1.5 - cellWidth / 2;
            self.pageL ++;
            [self.scrollView setContentOffset:CGPointMake(self.pointL, 0) animated:YES];
        } else if (self.pageL == adCount){
            self.pointL = 0;
            self.pageL = 1;
            [self.scrollView setContentOffset:CGPointMake(0, 0) animated:NO];
        } else {
            self.pointL += self.adLandscapeWidth;
            self.pageL ++;
            [self.scrollView setContentOffset:CGPointMake(self.pointL, 0) animated:YES];
        }
    }
    
}

#pragma mark - NADNativeDelegate

- (void)nadNativeDidClickAd:(NADNative *)ad
{
    NSLog(@"nadNativeDidClickAd: %@", ad);
}

@end

File: Sample/NendSDK_Swift_Sample/NativeAdCarouselCell.swift
//
//  NativeAdCarouselCell.swift
//  NendSDK_Sample
//
//  Copyright © 2016年 F@N Communications. All rights reserved.
//

import UIKit

func CreateTimer(interval: Double, queue: dispatch_queue_t, block: dispatch_block_t) -> dispatch_source_t
{
    let timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue)
    dispatch_source_set_timer(timer, dispatch_time(DISPATCH_TIME_NOW, Int64(timerInterval * Double(NSEC_PER_SEC))), UInt64(timerInterval * Double(NSEC_PER_SEC)), (1 * NSEC_PER_SEC) / 10)
    dispatch_source_set_event_handler(timer, block);
    dispatch_resume(timer)
    
    return timer;
}

let adCount: Int = 5
let timerInterval: NSTimeInterval = 3.0

class NativeAdCarouselCell: UITableViewCell, UIScrollViewDelegate, NADNativeDelegate {
    
    internal var direction: Int = 0
    
    private var client: NADNativeClient!
    private var adViewsP = [NativeAdCarouselView]()
    private var adViewsL = [NativeAdCarouselView]()
    private var ads = [NADNative]()
    private var scrollView: UIScrollView!
    private var pointP: CGFloat!
    private var pageP: CGFloat!
    private var pointL: CGFloat!
    private var pageL: CGFloat!
    
    private var timerP: dispatch_source_t!
    private var timerL: dispatch_source_t!
    
    private var landscapeWidth: CGFloat!
    private var cellWidth: CGFloat!
    
    deinit {
        print("NativeAdCarouselCell: deinit")
    }
    
    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        self.scrollView = UIScrollView()
        self.scrollView.delegate = self
        self.scrollView.backgroundColor = UIColor.clearColor()
        self.scrollView.showsHorizontalScrollIndicator = false
        self.scrollView.directionalLockEnabled = true
        self.scrollView.pagingEnabled = false
        self.scrollView.bounces = false
        self.scrollView.decelerationRate = 0.3
        self.addSubview(self.scrollView)
        NSNotificationCenter.defaultCenter().addObserver(self, selector:#selector(NativeAdCarouselCell.layoutUpdate(_:)), name: "layoutUpdate", object: nil)
        
        // 自動スクロール
        self.pointP = 0
        self.pageP = 1
        self.pointL = 0
        self.pageL = 1
    }
    
    func initAd() {
        self.client = NADNativeClient(spotId: "485504", apiKey: "30fda4b3386e793a14b27bedb4dcd29f03d638e5")
        NADNativeLogger.setLogLevel(.Debug)
        
        UIApplication.sharedApplication().networkActivityIndicatorVisible = true
        let group = dispatch_group_create()
        dispatch_apply(adCount, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)) { (i) -> Void in
            dispatch_group_enter(group)
            self.client.loadWithCompletionBlock({ (ad, _) -> Void in
                if ad != nil {
                    let nibP = UINib(nibName: "NativeAdCarouselPortraitView", bundle: nil)
                    let nibL = UINib(nibName: "NativeAdCarouselLandscapeView", bundle: nil)
                    let viewP: NativeAdCarouselView = nibP.instantiateWithOwner(self, options: nil)[0] as! NativeAdCarouselView
                    let viewL: NativeAdCarouselView = nibL.instantiateWithOwner(self, options: nil)[0] as! NativeAdCarouselView
                    
                    self.ads.append(ad)
                    self.adViewsP.append(viewP)
                    self.adViewsL.append(viewL)
                }
                dispatch_group_leave(group)
            })
        }
        dispatch_group_notify(group, dispatch_get_main_queue()) { () -> Void in
            UIApplication.sharedApplication().networkActivityIndicatorVisible = false
            self.setAd()
        }
    }
    
    func setAd() {
        if self.direction == 1 {
            for i in 0 ..< self.adViewsP.count {
                let viewP: NativeAdCarouselView = (self.adViewsP)[i]
                viewP.frame = CGRectMake(CGFloat(i) * adPortraitWidth, 0, adPortraitWidth, adPortraitHeight)
                
                dispatch_async(dispatch_get_main_queue(), {
                    let ad: NADNative = self.ads[i]
                    ad.delegate = self;
                    ad.intoView(viewP, advertisingExplicitly: .Promotion)
                })
            }
        } else if self.direction == 2 {
            for i in 0 ..< self.adViewsL.count {
                let viewL: NativeAdCarouselView = (self.adViewsL)[i]
                viewL.frame = CGRectMake(CGFloat(i) * adPortraitWidth, 0, adPortraitWidth, adPortraitHeight)
                
                dispatch_async(dispatch_get_main_queue(), {
                    let ad: NADNative = self.ads[i]
                    ad.delegate = self;
                    ad.intoView(viewL, advertisingExplicitly: .Promotion)
                })
            }
        }
        
        // スクロールビューに広告ビューを格納
        self.layoutUpdate()
    }
    
    func layoutUpdate(notification: NSNotification) {
        self.direction = notification.object as! Int
        UIScreen.mainScreen().bounds.size.width
        
        for subview in self.scrollView.subviews {
            subview.removeFromSuperview()
        }
        
        dispatch_async(dispatch_get_main_queue(), {
            self.setAd()
        })
    }
    
    func layoutUpdate() {
        self.cellWidth = UIScreen.mainScreen().bounds.size.width
        var width: CGFloat
        var height: CGFloat
        
        if self.direction == 1 {
            self.landscapeWidth = adPortraitWidth
            width = self.landscapeWidth
            height = adPortraitHeight
            
            for i in 0 ..< self.adViewsP.count {
                let adViewP: NativeAdCarouselView = (self.adViewsP)[i] as NativeAdCarouselView
                adViewP.index = i
                adViewP.frameUpdate(self.direction)
                self.scrollView.addSubview(adViewP)
            }
            self.scrollView.setContentOffset(CGPointMake(self.pointP, 0), animated:true)
        } else {
            if (self.cellWidth < adLandscapeWidth) {
                self.landscapeWidth = self.cellWidth - 20
            } else {
                self.landscapeWidth = adLandscapeWidth
            }
            width = self.landscapeWidth
            height = adLandscapeHeight
            
            for i in 0 ..< self.adViewsL.count {
                let adViewL: NativeAdCarouselView = (self.adViewsL)[i] as NativeAdCarouselView
                adViewL.index = i
                adViewL.frameUpdate(self.direction)
                self.scrollView.addSubview(adViewL)
            }
            self.scrollView.setContentOffset(CGPointMake(self.pointL, 0), animated:true)
        }
        
        self.scrollView.frame = CGRectMake(0, 0, self.cellWidth, height)
        self.scrollView.contentSize = CGSizeMake(width * CGFloat(adCount), 0)
        
        self.pageP = 1
        self.pageL = 1
        self.pointP = 0
        self.pointL = 0
        self.scrollView.setContentOffset(CGPointMake(0, 0), animated:false)
        // 自動スクロールタイマー設置
        if adCount > 1 {
            self.setTimer()
        }
    }
    
    // MARK: - UIScrollViewDelegate
    func scrollViewDidEndDragging(scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        self.animation()
    }
    
    func scrollViewDidEndDecelerating(scrollView: UIScrollView) {
        self.animation()
    }
    
    func animation() {
        var distance: CGFloat = self.scrollView.contentOffset.x
        var adWidth: CGFloat
        if (self.direction == 1) {
            adWidth = adPortraitWidth
            
            for i in 1 ..< adCount + 1 {
                if distance + self.cellWidth / 2 < adWidth * CGFloat(i) {
                    if (i == 1) {
                        distance = 0
                    } else if (i == adCount) {
                        distance = adWidth * CGFloat(i) - self.cellWidth
                    } else {
                        distance = adWidth * CGFloat(i) - adWidth / 2 - self.cellWidth / 2
                    }
                    self.pageP = CGFloat(i)
                    break
                }
            }
            
            self.pointP = distance
        } else {
            adWidth = self.landscapeWidth
            
            for i in 1 ..< adCount + 1 {
                if distance + self.cellWidth / 2 < adWidth * CGFloat(i) {
                    if (i == 1) {
                        distance = 0
                    } else if (i == adCount) {
                        distance = adWidth * CGFloat(i) - self.cellWidth
                    } else {
                        distance = adWidth * CGFloat(i) - adWidth / 2 - self.cellWidth / 2
                    }
                    self.pageL = CGFloat(i)
                    break
                }
            }
            
            self.pointL = distance
        }
        
        self.scrollView.setContentOffset(CGPointMake(distance, 0), animated:true)
    }
    
    // 自動スクロール
    func setTimer() {
        self.cancelTimer()
        
        weak var weakSelf = self
        let queue = dispatch_get_main_queue()
        if self.direction == 1 {
            self.timerP = CreateTimer(timerInterval, queue: queue, block: {
                weakSelf!.move()
            })
        } else if self.direction == 2 {
            self.timerL = CreateTimer(timerInterval, queue: queue, block: {
                weakSelf!.move()
            })
        }
    }
    
    func cancelTimer() {
        if(self.timerP != nil) {
            dispatch_source_cancel(self.timerP)
            self.timerP = nil
        }
        if(self.timerL != nil) {
            dispatch_source_cancel(self.timerL)
            self.timerL = nil
        }
    }
    
    func move() {
        if self.direction == 1 {
            if (self.pageP == CGFloat(adCount - 1)){
                self.pointP = adPortraitWidth * CGFloat(adCount) - cellWidth
                self.pageP = self.pageP + 1
                self.scrollView.setContentOffset(CGPointMake(self.pointP, 0), animated:true)
            } else if (self.pageP == 1) {
                self.pointP = self.pointP + adPortraitWidth * 1.5 - cellWidth / 2
                self.pageP = self.pageP + 1
                self.scrollView.setContentOffset(CGPointMake(self.pointP, 0), animated:true)
            } else if (self.pageP == CGFloat(adCount)){
                self.pointP = 0
                self.pageP = 1
                self.scrollView.setContentOffset(CGPointMake(0, 0), animated:false)
            } else {
                self.pointP = self.pointP + adPortraitWidth
                self.pageP = self.pageP + 1
                self.scrollView.setContentOffset(CGPointMake(self.pointP, 0), animated:true)
            }
            
        } else if self.direction == 2 {
            if (self.pageL == CGFloat(adCount - 1)){
                self.pointL = self.landscapeWidth * CGFloat(adCount) - cellWidth
                self.pageL = self.pageL + 1
                self.scrollView.setContentOffset(CGPointMake(self.pointL, 0), animated:true)
            } else if (self.pageL == 1) {
                self.pointL = self.pointL + self.landscapeWidth * 1.5 - cellWidth / 2
                self.pageL = self.pageL + 1
                self.scrollView.setContentOffset(CGPointMake(self.pointL, 0), animated:true)
            } else if (self.pageL == CGFloat(adCount)){
                self.pointL = 0
                self.pageL = 1
                self.scrollView.setContentOffset(CGPointMake(0, 0), animated:false)
            } else {
                self.pointL = self.pointL + self.landscapeWidth
                self.pageL = self.pageL + 1
                self.scrollView.setContentOffset(CGPointMake(self.pointL, 0), animated:true)
            }
        }
    }
    
    // MARK: - NADNativeDelegate
    func nadNativeDidClickAd(ad: NADNative!) {
        print(#function)
    }
}
